File: testData/java/UpperRules.expected.java
// ---- GeneratedParser.java -----------------
//header.txt
package generated;

import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiBuilder.Marker;
import static generated.GeneratedTypes.*;
import static com.intellij.lang.parser.GeneratedParserUtilBase.*;
import com.intellij.psi.tree.IElementType;
import com.intellij.lang.ASTNode;
import com.intellij.psi.tree.TokenSet;
import com.intellij.lang.PsiParser;
import com.intellij.lang.LightPsiParser;

@SuppressWarnings({"SimplifiableIfStatement", "UnusedAssignment"})
public class GeneratedParser implements PsiParser, LightPsiParser {

  public ASTNode parse(IElementType root_, PsiBuilder builder_) {
    parseLight(root_, builder_);
    return builder_.getTreeBuilt();
  }

  public void parseLight(IElementType root_, PsiBuilder builder_) {
    boolean result_;
    builder_ = adapt_builder_(root_, builder_, this, EXTENDS_SETS_);
    Marker marker_ = enter_section_(builder_, 0, _COLLAPSE_, null);
    result_ = parse_root_(root_, builder_);
    exit_section_(builder_, 0, marker_, root_, result_, true, TRUE_CONDITION);
  }

  protected boolean parse_root_(IElementType root_, PsiBuilder builder_) {
    return parse_root_(root_, builder_, 0);
  }

  static boolean parse_root_(IElementType root_, PsiBuilder builder_, int level_) {
    return file(builder_, level_ + 1);
  }

  public static final TokenSet[] EXTENDS_SETS_ = new TokenSet[] {
    create_token_set_(ABC, ABC_ONE, ABC_THREE, ABC_TWO),
  };

  /* ********************************************************** */
  // abc_three? (abc_one | abc_two | abc_three)
  public static boolean abc(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "abc")) return false;
    boolean result_;
    Marker marker_ = enter_section_(builder_, level_, _COLLAPSE_, ABC, "<abc>");
    result_ = abc_0(builder_, level_ + 1);
    result_ = result_ && abc_1(builder_, level_ + 1);
    exit_section_(builder_, level_, marker_, result_, false, null);
    return result_;
  }

  // abc_three?
  private static boolean abc_0(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "abc_0")) return false;
    abc_three(builder_, level_ + 1);
    return true;
  }

  // abc_one | abc_two | abc_three
  private static boolean abc_1(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "abc_1")) return false;
    boolean result_;
    result_ = abc_one(builder_, level_ + 1);
    if (!result_) result_ = abc_two(builder_, level_ + 1);
    if (!result_) result_ = abc_three(builder_, level_ + 1);
    return result_;
  }

  /* ********************************************************** */
  // just_b X
  public static boolean abc_one(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "abc_one")) return false;
    if (!nextTokenIs(builder_, B)) return false;
    boolean result_, pinned_;
    Marker marker_ = enter_section_(builder_, level_, _UPPER_, ABC_ONE, null);
    result_ = just_b(builder_, level_ + 1);
    pinned_ = result_; // pin = 1
    result_ = result_ && consumeToken(builder_, X);
    exit_section_(builder_, level_, marker_, result_, pinned_, null);
    return result_ || pinned_;
  }

  /* ********************************************************** */
  // X
  public static boolean abc_three(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "abc_three")) return false;
    if (!nextTokenIs(builder_, X)) return false;
    boolean result_;
    Marker marker_ = enter_section_(builder_);
    result_ = consumeToken(builder_, X);
    exit_section_(builder_, marker_, ABC_THREE, result_);
    return result_;
  }

  /* ********************************************************** */
  // C
  public static boolean abc_two(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "abc_two")) return false;
    if (!nextTokenIs(builder_, C)) return false;
    boolean result_;
    Marker marker_ = enter_section_(builder_, level_, _UPPER_, ABC_TWO, null);
    result_ = consumeToken(builder_, C);
    exit_section_(builder_, level_, marker_, result_, false, null);
    return result_;
  }

  /* ********************************************************** */
  // abc | pinned_seq | plain_seq
  static boolean content(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "content")) return false;
    boolean result_;
    Marker marker_ = enter_section_(builder_, level_, _NONE_);
    result_ = abc(builder_, level_ + 1);
    if (!result_) result_ = pinned_seq(builder_, level_ + 1);
    if (!result_) result_ = plain_seq(builder_, level_ + 1);
    exit_section_(builder_, level_, marker_, result_, false, GeneratedParser::content_recover);
    return result_;
  }

  /* ********************************************************** */
  // !';'
  static boolean content_recover(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "content_recover")) return false;
    boolean result_;
    Marker marker_ = enter_section_(builder_, level_, _NOT_);
    result_ = !consumeToken(builder_, ";");
    exit_section_(builder_, level_, marker_, result_, false, null);
    return result_;
  }

  /* ********************************************************** */
  // (content ';') *
  static boolean file(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "file")) return false;
    while (true) {
      int pos_ = current_position_(builder_);
      if (!file_0(builder_, level_ + 1)) break;
      if (!empty_element_parsed_guard_(builder_, "file", pos_)) break;
    }
    return true;
  }

  // content ';'
  private static boolean file_0(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "file_0")) return false;
    boolean result_, pinned_;
    Marker marker_ = enter_section_(builder_, level_, _NONE_);
    result_ = content(builder_, level_ + 1);
    pinned_ = result_; // pin = 1
    result_ = result_ && consumeToken(builder_, ";");
    exit_section_(builder_, level_, marker_, result_, pinned_, null);
    return result_ || pinned_;
  }

  /* ********************************************************** */
  // B
  public static boolean just_b(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "just_b")) return false;
    if (!nextTokenIs(builder_, B)) return false;
    boolean result_;
    Marker marker_ = enter_section_(builder_);
    result_ = consumeToken(builder_, B);
    exit_section_(builder_, marker_, JUST_B, result_);
    return result_;
  }

  /* ********************************************************** */
  // prefix (abc_one | abc_two)
  public static boolean pinned_seq(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "pinned_seq")) return false;
    if (!nextTokenIs(builder_, A)) return false;
    boolean result_, pinned_;
    Marker marker_ = enter_section_(builder_, level_, _NONE_, PINNED_SEQ, null);
    result_ = prefix(builder_, level_ + 1);
    pinned_ = result_; // pin = 1
    result_ = result_ && pinned_seq_1(builder_, level_ + 1);
    exit_section_(builder_, level_, marker_, result_, pinned_, null);
    return result_ || pinned_;
  }

  // abc_one | abc_two
  private static boolean pinned_seq_1(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "pinned_seq_1")) return false;
    boolean result_;
    result_ = abc_one(builder_, level_ + 1);
    if (!result_) result_ = abc_two(builder_, level_ + 1);
    return result_;
  }

  /* ********************************************************** */
  // prefix abc_one abc_two
  public static boolean plain_seq(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "plain_seq")) return false;
    if (!nextTokenIs(builder_, A)) return false;
    boolean result_;
    Marker marker_ = enter_section_(builder_);
    result_ = prefix(builder_, level_ + 1);
    result_ = result_ && abc_one(builder_, level_ + 1);
    result_ = result_ && abc_two(builder_, level_ + 1);
    exit_section_(builder_, marker_, PLAIN_SEQ, result_);
    return result_;
  }

  /* ********************************************************** */
  // A
  public static boolean prefix(PsiBuilder builder_, int level_) {
    if (!recursion_guard_(builder_, level_, "prefix")) return false;
    if (!nextTokenIs(builder_, A)) return false;
    boolean result_;
    Marker marker_ = enter_section_(builder_);
    result_ = consumeToken(builder_, A);
    exit_section_(builder_, marker_, PREFIX, result_);
    return result_;
  }

}

File: testData/kotlin/UpperRules.expected.kt
// ---- generated/GeneratedParser.kt -----------------
//header.txt
package generated

import com.intellij.platform.syntax.util.runtime.*
import com.intellij.platform.syntax.parser.SyntaxTreeBuilder.Marker
import com.intellij.platform.syntax.SyntaxElementTypeSet
import com.intellij.platform.syntax.syntaxElementTypeSetOf
import com.intellij.platform.syntax.SyntaxElementType

@Suppress("unused", "FunctionName", "JoinDeclarationAndAssignment")
object GeneratedParser {

  fun parse(root_: SyntaxElementType, runtime_: SyntaxGeneratedParserRuntime) {
    var result_: Boolean
    runtime_.init(::parse, EXTENDS_SETS_)
    val marker_: Marker = runtime_.enter_section_(0, Modifiers._COLLAPSE_, null)
    result_ = parse_root_(root_, runtime_, 0)
    runtime_.exit_section_(0, marker_, root_, result_, true, TRUE_CONDITION)
  }

  internal fun parse_root_(root_: SyntaxElementType, runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    return `file_$`(runtime_, level_ + 1)
  }

  val EXTENDS_SETS_: Array<SyntaxElementTypeSet> = arrayOf(
    create_token_set_(GeneratedSyntaxElementTypes.ABC, GeneratedSyntaxElementTypes.ABC_ONE, GeneratedSyntaxElementTypes.ABC_THREE, GeneratedSyntaxElementTypes.ABC_TWO),
  )

  /* ********************************************************** */
  // abc_three? (abc_one | abc_two | abc_three)
  fun abc(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "abc")) return false
    var result_: Boolean
    val marker_: Marker = runtime_.enter_section_(level_, Modifiers._COLLAPSE_, GeneratedSyntaxElementTypes.ABC, "<abc>")
    result_ = abc_0(runtime_, level_ + 1)
    result_ = result_ && abc_1(runtime_, level_ + 1)
    runtime_.exit_section_(level_, marker_, result_, false, null)
    return result_
  }

  // abc_three?
  private fun abc_0(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "abc_0")) return false
    abc_three(runtime_, level_ + 1)
    return true
  }

  // abc_one | abc_two | abc_three
  private fun abc_1(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "abc_1")) return false
    var result_: Boolean
    result_ = abc_one(runtime_, level_ + 1)
    if (!result_) result_ = abc_two(runtime_, level_ + 1)
    if (!result_) result_ = abc_three(runtime_, level_ + 1)
    return result_
  }

  /* ********************************************************** */
  // just_b X
  fun abc_one(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "abc_one")) return false
    if (!runtime_.nextTokenIs(GeneratedSyntaxElementTypes.B)) return false
    var result_: Boolean
    var pinned_: Boolean
    val marker_: Marker = runtime_.enter_section_(level_, Modifiers._UPPER_, GeneratedSyntaxElementTypes.ABC_ONE, null)
    result_ = just_b(runtime_, level_ + 1)
    pinned_ = result_ // pin = 1
    result_ = result_ && runtime_.consumeToken(GeneratedSyntaxElementTypes.X)
    runtime_.exit_section_(level_, marker_, result_, pinned_, null)
    return result_ || pinned_
  }

  /* ********************************************************** */
  // X
  fun abc_three(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "abc_three")) return false
    if (!runtime_.nextTokenIs(GeneratedSyntaxElementTypes.X)) return false
    var result_: Boolean
    val marker_: Marker = runtime_.enter_section_()
    result_ = runtime_.consumeToken(GeneratedSyntaxElementTypes.X)
    runtime_.exit_section_(marker_, GeneratedSyntaxElementTypes.ABC_THREE, result_)
    return result_
  }

  /* ********************************************************** */
  // C
  fun abc_two(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "abc_two")) return false
    if (!runtime_.nextTokenIs(GeneratedSyntaxElementTypes.C)) return false
    var result_: Boolean
    val marker_: Marker = runtime_.enter_section_(level_, Modifiers._UPPER_, GeneratedSyntaxElementTypes.ABC_TWO, null)
    result_ = runtime_.consumeToken(GeneratedSyntaxElementTypes.C)
    runtime_.exit_section_(level_, marker_, result_, false, null)
    return result_
  }

  /* ********************************************************** */
  // abc | pinned_seq | plain_seq
  internal fun content(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "content")) return false
    var result_: Boolean
    val marker_: Marker = runtime_.enter_section_(level_, Modifiers._NONE_)
    result_ = abc(runtime_, level_ + 1)
    if (!result_) result_ = pinned_seq(runtime_, level_ + 1)
    if (!result_) result_ = plain_seq(runtime_, level_ + 1)
    runtime_.exit_section_(level_, marker_, result_, false, GeneratedParser::content_recover)
    return result_
  }

  /* ********************************************************** */
  // !';'
  internal fun content_recover(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "content_recover")) return false
    var result_: Boolean
    val marker_: Marker = runtime_.enter_section_(level_, Modifiers._NOT_)
    result_ = !runtime_.consumeToken(";")
    runtime_.exit_section_(level_, marker_, result_, false, null)
    return result_
  }

  /* ********************************************************** */
  // (content ';') *
  internal fun `file_$`(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "`file_$`")) return false
    while (true) {
      val pos_: Int = runtime_.current_position_()
      if (!file_0(runtime_, level_ + 1)) break
      if (!runtime_.empty_element_parsed_guard_("`file_$`", pos_)) break
    }
    return true
  }

  // content ';'
  private fun file_0(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "file_0")) return false
    var result_: Boolean
    var pinned_: Boolean
    val marker_: Marker = runtime_.enter_section_(level_, Modifiers._NONE_)
    result_ = content(runtime_, level_ + 1)
    pinned_ = result_ // pin = 1
    result_ = result_ && runtime_.consumeToken(";")
    runtime_.exit_section_(level_, marker_, result_, pinned_, null)
    return result_ || pinned_
  }

  /* ********************************************************** */
  // B
  fun just_b(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "just_b")) return false
    if (!runtime_.nextTokenIs(GeneratedSyntaxElementTypes.B)) return false
    var result_: Boolean
    val marker_: Marker = runtime_.enter_section_()
    result_ = runtime_.consumeToken(GeneratedSyntaxElementTypes.B)
    runtime_.exit_section_(marker_, GeneratedSyntaxElementTypes.JUST_B, result_)
    return result_
  }

  /* ********************************************************** */
  // prefix (abc_one | abc_two)
  fun pinned_seq(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "pinned_seq")) return false
    if (!runtime_.nextTokenIs(GeneratedSyntaxElementTypes.A)) return false
    var result_: Boolean
    var pinned_: Boolean
    val marker_: Marker = runtime_.enter_section_(level_, Modifiers._NONE_, GeneratedSyntaxElementTypes.PINNED_SEQ, null)
    result_ = prefix(runtime_, level_ + 1)
    pinned_ = result_ // pin = 1
    result_ = result_ && pinned_seq_1(runtime_, level_ + 1)
    runtime_.exit_section_(level_, marker_, result_, pinned_, null)
    return result_ || pinned_
  }

  // abc_one | abc_two
  private fun pinned_seq_1(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "pinned_seq_1")) return false
    var result_: Boolean
    result_ = abc_one(runtime_, level_ + 1)
    if (!result_) result_ = abc_two(runtime_, level_ + 1)
    return result_
  }

  /* ********************************************************** */
  // prefix abc_one abc_two
  fun plain_seq(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "plain_seq")) return false
    if (!runtime_.nextTokenIs(GeneratedSyntaxElementTypes.A)) return false
    var result_: Boolean
    val marker_: Marker = runtime_.enter_section_()
    result_ = prefix(runtime_, level_ + 1)
    result_ = result_ && abc_one(runtime_, level_ + 1)
    result_ = result_ && abc_two(runtime_, level_ + 1)
    runtime_.exit_section_(marker_, GeneratedSyntaxElementTypes.PLAIN_SEQ, result_)
    return result_
  }

  /* ********************************************************** */
  // A
  fun prefix(runtime_: SyntaxGeneratedParserRuntime, level_: Int): Boolean {
    if (!runtime_.recursion_guard_(level_, "prefix")) return false
    if (!runtime_.nextTokenIs(GeneratedSyntaxElementTypes.A)) return false
    var result_: Boolean
    val marker_: Marker = runtime_.enter_section_()
    result_ = runtime_.consumeToken(GeneratedSyntaxElementTypes.A)
    runtime_.exit_section_(marker_, GeneratedSyntaxElementTypes.PREFIX, result_)
    return result_
  }

}